Enable JIT tests in fbcode

zdevito opened the pull request:
Summary: Enables JIT tests in FBCode. Changes pybind11 code to avoid mixing py::args with positinally matched arguments because old versions of PyBind11 leak memory in this case.

Differential Revision: D10419708

apaszke commented on test/test_jit.py:
```diff
@@ -640,7 +640,7 @@ def scaleshift(x, scale, shift):
         ]
         ge = self.checkTrace(scaleshift, inputs)
-        self.assertExpectedGraph(ge.graph_for(*inputs))
+        # self.assertExpectedGraph(ge.graph_for(*inputs))
```
Why is this commented out?

zdevito replied:
because in phabricator the diff is covered with linter errors, so I can't actually see what I am doing :)

apaszke commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -250,31 +250,29 @@ inline py::object toPyObject(IValue&& ivalue) {
   // has one.
   size_t consumed_kwargs = 0;
-  for (size_t i = args.size(); i < schema.arguments.size(); ++i) {
+  for (size_t i = args_size; i < schema.arguments.size(); ++i) {
```
Ugh `args_size` and `args_start` are so easy to confuse in here 😕

apaszke commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -250,31 +250,29 @@ inline py::object toPyObject(IValue&& ivalue) {
 inline Stack createStackForSchema(
     const FunctionSchema& schema,
     py::args args,
+    size_t arg_start,
```
This is soooo ugly 😞

zdevito replied:
Agreed, but the alternatives are worse:

1. slicing the tuple requires allocating a new tuple, which is added work along the invocation pathway. 
2. upgrading pybind11 in tp2 is really tedious.

apaszke replied:
I agree we don't want to allocate a temporary in there. It's just pretty sad that some random fbcode requirements are forcing us to make the code this much worse 😕

apaszke reviewed (dismissed):
You have probably accidentally commented out a line in the JIT tests + all `pybind` places that you've changed should have a comment, or we'll inadvertently revert them to what we had today (which looks *way* better).

apaszke commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -247,19 +255,40 @@ inline py::object toPyObject(IValue&& ivalue) {
   }
 }
 
+struct VISIBILITY_HIDDEN tuple_slice {
```
That's great!

zdevito committed:
Enable JIT tests in fbcode (#12777)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12777

Enables JIT tests in FBCode. Changes pybind11 code to avoid mixing py::args with positinally matched arguments because old versions of PyBind11 leak memory in this case.

Reviewed By: jamesr66a

Differential Revision: D10419708

fbshipit-source-id: 5f70be006de95ec3aa73ec1d9ff3e15b00f2d811
